pkg/cmd/admin/migrate/images/images.go

```diff
@@ -0,0 +1,488 @@
+	// any object with a pod template
+	return nil, nil
+}
+
+// reporter implements the Reporter interface for a boolean.
+type reporter bool
+
+func (r reporter) Changed() bool {
+	return bool(r)
+}
+
+// imageChangeInfo indicates whether the spec or status of an image stream was changed.
+type imageChangeInfo struct {
+	spec, status bool
+}
+
+func (i imageChangeInfo) Changed() bool {
+	return i.spec || i.status
+}
+
+type TransformImageFunc func(in string) string
+
+func updateString(value *string, fn TransformImageFunc) bool {
+	result := fn(*value)
+	if result != *value {
+		*value = result
+		return true
+	}
+	return false
+}
+
+func updatePodSpec(spec *kapi.PodSpec, fn TransformImageFunc) bool {
+	var changed bool
+	for i := range spec.Containers {
+		changed = updateString(&spec.Containers[i].Image, fn) || changed
+	}
+	return changed
+}
+
+func updatePodTemplate(template *kapi.PodTemplateSpec, fn TransformImageFunc) bool {
+	if template == nil {
+		return false
+	}
+	return updatePodSpec(&template.Spec, fn)
+}
+
+func updateDockerConfig(cfg credentialprovider.DockerConfig, fn TransformImageFunc) bool {
+	var changed bool
+	for k, v := range cfg {
+		original := k
+		if updateString(&k, fn) {
+			changed = true
+			delete(cfg, original)
+			cfg[k] = v
+		}
+	}
+	return changed
+}
+
+// ImageReferenceMapping represents a transformation of an image reference.
+type ImageReferenceMapping struct {
+	FromRegistry string
+	FromName     string
+	ToRegistry   string
+	ToName       string
+}
+
+// ParseMapping converts a string in the form "(REGISTRY|*)/(NAME|*)" to an ImageReferenceMapping
+// or returns a user-facing error. REGISTRY is the image registry value (hostname) or "docker.io".
+// NAME is the full repository name (the path relative to the registry root).
+// TODO: handle v2 repository names, which can have multiple segments (must fix
+//   ParseDockerImageReference)
+func ParseMapping(s string) (ImageReferenceMapping, error) {
+	parts := strings.SplitN(s, "=", 2)
+	from := strings.SplitN(parts[0], "/", 2)
+	to := strings.SplitN(parts[1], "/", 2)
+	if len(from) < 2 || len(to) < 2 {
+		return ImageReferenceMapping{}, fmt.Errorf("all arguments must be of the form REGISTRY/NAME=REGISTRY/NAME, where registry or name may be '*' or a value")
+	}
+	if len(from[0]) == 0 {
+		return ImageReferenceMapping{}, fmt.Errorf("%q is not a valid source: registry must be specified (may be '*')", parts[0])
+	}
+	if len(from[1]) == 0 {
+		return ImageReferenceMapping{}, fmt.Errorf("%q is not a valid source: name must be specified (may be '*')", parts[0])
+	}
+	if len(to[0]) == 0 {
+		return ImageReferenceMapping{}, fmt.Errorf("%q is not a valid target: registry must be specified (may be '*')", parts[1])
+	}
+	if len(to[1]) == 0 {
+		return ImageReferenceMapping{}, fmt.Errorf("%q is not a valid target: name must be specified (may be '*')", parts[1])
+	}
+	if from[0] == "*" {
+		from[0] = ""
```

Comment:
Why shouldn't `*` mean "all" for from?  I haven't regretted forcing that behavior in our policy API, but I absolutely hate the different handling of `namespace == ""` in our API.